### PR TITLE
[SPARK-46394][SQL] Fix spark.catalog.listDatabases() issues on schemas with special characters when `spark.sql.legacy.keepCommandOutputSchema` set to true 

### PR DESCRIPTION
**Pull request by anchovYu**
### What changes were proposed in this pull request?
When the SQL conf `spark.sql.legacy.keepCommandOutputSchema` is set to true:
Before:
```
// support there is a xyyu-db-with-hyphen schema in the catalog
spark.catalog.listDatabases()

[INVALID_IDENTIFIER] The identifier xyyu-db-with-hyphen is invalid. Please, consider quoting it with back-quotes as `xyyu-db-with-hyphen`. SQLSTATE: 42602 (line 1, pos 4)
```

After:
```
spark.catalog.listDatabases()

.. `xyyu-db-with-hyphen` ..
```

This PR fixes the issue by falling back to original name when the parsing failed.

### Why are the changes needed?
To fix the bug.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Newly added tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 1]
```diff
@@ -100,7 +109,9 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
       case ShowNamespaces(r: ResolvedNamespace, _, _) => r.catalog
     }.get
     val databases = qe.toRdd.collect().map { row =>
-      makeDatabase(Some(catalog.name()), row.getString(0))
+      // dbName can either be a quoted identifier (single or multi part) or an unquoted single part
+      val dbName = row.getString(0)
+      makeDatabase(Some(catalog.name()), dbName)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:114]: seems unnecessary change?

Comment (anchovYu) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala:114]: Yes, but it's easier to reference in the comment (`dbName` instead of something like "returned results" or `row.getString(0)`.

File: sql/core/src/test/scala/org/apache/spark/sql/internal/CatalogSuite.scala [hunk 2]
```diff
@@ -167,6 +167,44 @@ class CatalogSuite extends SharedSparkSession with AnalysisTest with BeforeAndAf
...
+        assert(spark.catalog.listDatabases().collect().map(_.name).toSet ==
+          Set("default", "`my-db1`", "`my``db2`"))
+        if (legacy) {
+          assert(
+            spark.catalog.listDatabases("my*").collect().map(_.name).toSet ==
+              Set("`my-db1`", "`my``db2`")
+          )
+          assert(spark.catalog.listDatabases("`my*`").collect().map(_.name).toSet == Set.empty)
+        } else {
+          assert(spark.catalog.listDatabases("my*").collect().map(_.name).toSet == Set.empty)
```
[the remark this answers is not in the document]
Comment (anchovYu) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/CatalogSuite.scala:187]: It's the existing behavior in SQL : ).
With this conf flipped,  sql query `SHOW SCHEMAS LIKE` actually requires different pattern matching. This is because it does the quoting before pattern matching in the `ShowNamespaceExec`.

[the remark this answers is not in the document]
Comment (anchovYu) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/CatalogSuite.scala:190]: Added a todo.

**Review by cloud-fan — APPROVED** [review]
